src/Site/Controller/Redirect.php

```diff
@@ -0,0 +1,85 @@
+<?php
+
+namespace Site\Controller;
+
+use Api\Library\Shared\Palaso\Exception\UserUnauthorizedException;
+use Api\Library\Shared\SilexSessionHelper;
+use Api\Model\Shared\ProjectModel;
+use Api\Model\Shared\UserModel;
+use Silex\Application;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+class Redirect extends App
+{
+    /**
+     * @param Request $request
+     * @param Application $app
+     * @param $appName
+     * @param string $projectId
+     * @return Response
+     */
+    public function view(
+        /** @noinspection PhpUnusedParameterInspection */
+        Request $request, Application $app, $appName, $projectId = ''
+    ) {
+        $model = new AppModel($app, $appName, $this->website, $projectId);
+        try {
+            $this->setupBaseVariables($app);
+        } catch (\Exception $e) {
+            return $app->redirect('/auth/logout');
+        }
+        try {
+            if ($model->requireProject) {
+                if ($model->isPublicApp) {
+                    $model->projectId = SilexSessionHelper::requireValidProjectIdForThisWebsite($model->app, $this->website, $model->projectId);
+                } else {
+                    $model->projectId =
+                        SilexSessionHelper::requireValidProjectIdForThisWebsiteAndValidateUserMembership($model->app, $this->website, $model->projectId);
+                }
+            }
+        } catch (UserUnauthorizedException $e) {
+            if (SilexSessionHelper::getUserId($app)) {
+                // User tried to access project they're not a member of, so show them projects view so they can pick a different one
+                return $app->redirect("/app/projects");
+            }
+            return $app->redirect('/auth/logout');
+        } catch (\Exception $e) {
+            return $app->redirect('/auth/logout');
+        }
+        if ($model->projectId) {
+            try {
+                $project = new ProjectModel($model->projectId);
+                if ($project && $project->appName) {
+                    return $app->redirect("/app/$project->appName/$model->projectId");
+                }
+            } catch (\Exception $e) {
+                return $app->redirect('/auth/logout');
+            }
+        }
+        if ($this->_user && $this->_user->projects) {
+            if (count($this->_user->projects->refs) == 1) {
```

Comment:
Where is the code for `if (count($this->_user->projects->refs) > 1)` ?

I expected to see a block of code for when the user has multiple projects.  In this case we should redirect to `$this->_user->lastUsedProjectId` .  I think we can keep this controller simple and let the `/app/lexicon` controller handle the case where somehow the user doesn't have permission on the project anymore (or it doesn't exist).

Reply:
It's the final fallthrough case. I've pushed a new commit with comments explaining the redirects, to make the logic a bit easier to understand. The general principle that the logic should be understandable by just reading code, and that comments are superfluous, doesn't apply here as the reasoning behind the redirects isn't visible in the code.